File: backend/app/evaluators/deterministic.py
"""Deterministic evaluators: pure functions of the sample, no model calls."""

import math
import re
from typing import Any, Literal

from jsonschema import Draft202012Validator
from jsonschema.exceptions import SchemaError
from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.evaluators.base import EvaluationSample, EvaluatorError, EvaluatorOutcome
from app.evaluators.output import (
    JsonParseError,
    PathNotFoundError,
    parse_json_output,
    resolve_path,
    to_text,
)


class _Config(BaseModel):
    model_config = ConfigDict(extra="forbid")


def _expected_value(sample: EvaluationSample, path: str | None) -> Any:
    if sample.expected is None:
        raise EvaluatorError("This evaluator needs an expected output, but the test case has none")
    if path is None:
        return sample.expected
    try:
        return resolve_path(sample.expected, path)
    except PathNotFoundError as exc:
        raise EvaluatorError(f"Expected output has no field '{path}'") from exc


def _normalise(text: str, *, case_sensitive: bool, collapse_whitespace: bool) -> str:
    if collapse_whitespace:
        text = " ".join(text.split())
    return text if case_sensitive else text.casefold()


# --- exact match ---------------------------------------------------------------------------


class ExactMatchConfig(_Config):
    expected_path: str | None = Field(
        default=None, description="Field of the expected output to compare against."
    )
    output_path: str | None = Field(
        default=None, description="If set, parse the output as JSON and compare this field."
    )
    case_sensitive: bool = True
    collapse_whitespace: bool = True


class ExactMatchEvaluator:
    def __init__(self, config: ExactMatchConfig) -> None:
        self.config = config

    async def evaluate(self, sample: EvaluationSample) -> EvaluatorOutcome:
        expected = _expected_value(sample, self.config.expected_path)
        actual: Any = sample.output
        if self.config.output_path is not None:
            try:
                actual = resolve_path(parse_json_output(sample.output), self.config.output_path)
            except (JsonParseError, PathNotFoundError):
                return EvaluatorOutcome.binary(
                    False, f"Output has no JSON field '{self.config.output_path}'"
                )

        if isinstance(expected, str) or isinstance(actual, str):
            options = {
                "case_sensitive": self.config.case_sensitive,
                "collapse_whitespace": self.config.collapse_whitespace,
            }
            matched = _normalise(to_text(actual), **options) == _normalise(
                to_text(expected), **options
            )
        else:
            matched = actual == expected
        if matched:
            return EvaluatorOutcome.binary(True, "Output exactly matches the expected value")
        return EvaluatorOutcome.binary(
            False,
            f"Expected {to_text(expected)!r} but got {_truncate(to_text(actual))!r}",
        )


# --- contains ------------------------------------------------------------------------------


class ContainsConfig(_Config):
    values: list[str] = Field(default_factory=list, description="Fixed substrings to look for.")
    expected_path: str | None = Field(
        default=None,
        description="Also look for the value(s) at this field of the expected output.",
    )
    mode: Literal["all", "any"] = "all"
    case_sensitive: bool = False


class ContainsEvaluator:
    def __init__(self, config: ContainsConfig) -> None:
        self.config = config

    def _needles(self, sample: EvaluationSample) -> list[str]:
        needles = list(self.config.values)
        if self.config.expected_path is not None or (not needles and sample.expected is not None):
            expected = _expected_value(sample, self.config.expected_path)
            items = expected if isinstance(expected, list) else [expected]
            needles.extend(to_text(item) for item in items)
        if not needles:
            raise EvaluatorError("No values configured and no expected output to search for")
        return needles

    async def evaluate(self, sample: EvaluationSample) -> EvaluatorOutcome:
        needles = self._needles(sample)
        haystack = _normalise(
            sample.output, case_sensitive=self.config.case_sensitive, collapse_whitespace=True
        )
        found = [
            n
            for n in needles
            if _normalise(n, case_sensitive=self.config.case_sensitive, collapse_whitespace=True)
            in haystack
        ]
        missing = [n for n in needles if n not in found]
        if self.config.mode == "any":
            passed = bool(found)
            score = 1.0 if passed else 0.0
        else:
            passed = not missing
            score = len(found) / len(needles)
        reason = f"Found {len(found)} of {len(needles)} expected values" + (
            f"; missing {', '.join(repr(m) for m in missing[:5])}" if missing else ""
        )
        return EvaluatorOutcome(
            score=score, passed=passed, reason=reason, details={"found": found, "missing": missing}
        )


# --- regex ---------------------------------------------------------------------------------


class RegexConfig(_Config):
    pattern: str
    should_match: bool = True
    ignore_case: bool = False
    multiline: bool = False
    dotall: bool = False

    @field_validator("pattern")
    @classmethod
    def _compiles(cls, value: str) -> str:
        try:
            re.compile(value)
        except re.error as exc:
            raise ValueError(f"Invalid regular expression: {exc}") from exc
        return value


class RegexEvaluator:
    def __init__(self, config: RegexConfig) -> None:
        self.config = config
        flags = (
            (re.IGNORECASE if config.ignore_case else 0)
            | (re.MULTILINE if config.multiline else 0)
            | (re.DOTALL if config.dotall else 0)
        )
        self._regex = re.compile(config.pattern, flags)

    async def evaluate(self, sample: EvaluationSample) -> EvaluatorOutcome:
        match = self._regex.search(sample.output)
        passed = (match is not None) == self.config.should_match
        if match is not None:
            reason = f"Pattern matched {match.group(0)!r}"
        else:
            reason = "Pattern did not match"
        if not self.config.should_match:
            reason += " (expected no match)"
        return EvaluatorOutcome.binary(passed, reason, match=match.group(0) if match else None)


# --- JSON validity and schema --------------------------------------------------------------


class JsonValidConfig(_Config):
    allow_code_fence: bool = Field(
        default=False, description="Accept JSON wrapped in a markdown ```json fence."
    )


class JsonValidEvaluator:
    def __init__(self, config: JsonValidConfig) -> None:
        self.config = config

    async def evaluate(self, sample: EvaluationSample) -> EvaluatorOutcome:
        try:
            parse_json_output(sample.output, allow_code_fence=self.config.allow_code_fence)
        except JsonParseError as exc:
            return EvaluatorOutcome.binary(False, str(exc))
        return EvaluatorOutcome.binary(True, "Output is valid JSON")


class JsonSchemaConfig(_Config):
    json_schema: dict[str, Any] = Field(description="JSON Schema (draft 2020-12).")
    allow_code_fence: bool = False

    @field_validator("json_schema")
    @classmethod
    def _valid_schema(cls, value: dict[str, Any]) -> dict[str, Any]:
        try:
            Draft202012Validator.check_schema(value)
        except SchemaError as exc:
            raise ValueError(f"Invalid JSON Schema: {exc.message}") from exc
        return value


class JsonSchemaEvaluator:
    def __init__(self, config: JsonSchemaConfig) -> None:
        self.config = config
        self._validator = Draft202012Validator(config.json_schema)

    async def evaluate(self, sample: EvaluationSample) -> EvaluatorOutcome:
        try:
            value = parse_json_output(sample.output, allow_code_fence=self.config.allow_code_fence)
        except JsonParseError as exc:
            return EvaluatorOutcome.binary(False, str(exc))
        errors = sorted(self._validator.iter_errors(value), key=lambda e: list(e.path))
        if not errors:
            return EvaluatorOutcome.binary(True, "Output conforms to the schema")
        messages = [
            f"{'/'.join(str(p) for p in e.path) or '(root)'}: {e.message}" for e in errors[:10]
        ]
        return EvaluatorOutcome.binary(
            False,
            f"{len(errors)} schema violation(s): {messages[0]}",
            violations=messages,
        )


# --- required fields -----------------------------------------------------------------------


class RequiredFieldsConfig(_Config):
    fields: list[str] = Field(min_length=1, description="Dotted paths that must be present.")
    allow_null: bool = False
    allow_code_fence: bool = False


class RequiredFieldsEvaluator:
    def __init__(self, config: RequiredFieldsConfig) -> None:
        self.config = config

    async def evaluate(self, sample: EvaluationSample) -> EvaluatorOutcome:
        try:
            value = parse_json_output(sample.output, allow_code_fence=self.config.allow_code_fence)
        except JsonParseError as exc:
            return EvaluatorOutcome(
                score=0.0,
                passed=False,
                reason=str(exc),
                details={"missing": self.config.fields},
            )
        missing: list[str] = []
        for path in self.config.fields:
            try:
                found = resolve_path(value, path)
            except PathNotFoundError:
                missing.append(path)
                continue
            if found is None and not self.config.allow_null:
                missing.append(path)
        present = len(self.config.fields) - len(missing)
        reason = (
            "All required fields are present"
            if not missing
            else f"Missing {len(missing)} of {len(self.config.fields)} fields: {', '.join(missing)}"
        )
        return EvaluatorOutcome(
            score=present / len(self.config.fields),
            passed=not missing,
            reason=reason,
            details={"missing": missing},
        )


# --- field-level structured comparison -----------------------------------------------------


class FieldMatchConfig(_Config):
    fields: list[str] | None = Field(
        default=None, description="Fields to compare. Defaults to every top-level expected key."
    )
    numeric_tolerance: float = Field(
        default=0.0, ge=0, description="Allowed relative difference for numbers (0.01 = 1%)."
    )
    case_sensitive: bool = False
    pass_threshold: float = Field(
        default=1.0, ge=0, le=1, description="Fraction of fields that must match to pass."
    )
    allow_code_fence: bool = False


class FieldMatchEvaluator:
    def __init__(self, config: FieldMatchConfig) -> None:
        self.config = config

    def _values_match(self, expected: Any, actual: Any) -> bool:
        if isinstance(expected, bool) or isinstance(actual, bool):
            return bool(expected == actual)
        if isinstance(expected, int | float) and isinstance(actual, int | float):
            if self.config.numeric_tolerance == 0:
                return math.isclose(expected, actual)
            return math.isclose(expected, actual, rel_tol=self.config.numeric_tolerance)
        if isinstance(expected, str) and isinstance(actual, str):
            return _normalise(
                expected, case_sensitive=self.config.case_sensitive, collapse_whitespace=True
            ) == _normalise(
                actual, case_sensitive=self.config.case_sensitive, collapse_whitespace=True
            )
        return bool(expected == actual)

    async def evaluate(self, sample: EvaluationSample) -> EvaluatorOutcome:
        expected = _expected_value(sample, None)
        fields = self.config.fields
        if fields is None:
            if not isinstance(expected, dict):
                raise EvaluatorError("Field match needs an expected JSON object or explicit fields")
            fields = list(expected.keys())
        if not fields:
            raise EvaluatorError("No fields to compare")

        try:
            actual = parse_json_output(sample.output, allow_code_fence=self.config.allow_code_fence)
        except JsonParseError as exc:
            return EvaluatorOutcome(score=0.0, passed=False, reason=str(exc), details={})

        comparisons: list[dict[str, Any]] = []
        for path in fields:
            try:
                expected_value = resolve_path(expected, path)
            except PathNotFoundError as exc:
                raise EvaluatorError(f"Expected output has no field '{path}'") from exc
            try:
                actual_value = resolve_path(actual, path)
                matched = self._values_match(expected_value, actual_value)
            except PathNotFoundError:
                actual_value, matched = None, False
            comparisons.append(
                {
                    "field": path,
                    "expected": expected_value,
                    "actual": actual_value,
                    "match": matched,
                }
            )

        matched_count = sum(1 for c in comparisons if c["match"])
        score = matched_count / len(comparisons)
        mismatched = [c["field"] for c in comparisons if not c["match"]]
        reason = (
            f"All {len(comparisons)} fields match"
            if not mismatched
            else f"{matched_count} of {len(comparisons)} fields match; "
            f"mismatched: {', '.join(mismatched)}"
        )
        return EvaluatorOutcome(
            score=score,
            passed=score >= self.config.pass_threshold,
            reason=reason,
            details={"fields": comparisons},
        )


def _truncate(text: str, limit: int = 120) -> str:
    return text if len(text) <= limit else text[: limit - 1] + "…"

File: frontend/src/demo/engine/evaluators/deterministic.ts
/** Port of `app/evaluators/deterministic.py`: pure checks, no model calls. */

import { isDict, pyCasefold, pyEq, pyIsClose, pyLen, pyRepr, pySlice, pySplit } from "../py";
import {
  binary,
  EvaluatorError,
  type EvaluationSample,
  type Evaluator,
  type EvaluatorOutcome,
} from "./base";
import { SchemaValidator } from "./jsonschema";
import { JsonParseError, parseJsonOutput, PathNotFoundError, resolvePath, toText } from "./output";

type Config = Record<string, unknown>;

function expectedValue(sample: EvaluationSample, path: string | null): unknown {
  if (sample.expected === null || sample.expected === undefined) {
    throw new EvaluatorError("This evaluator needs an expected output, but the test case has none");
  }
  if (path === null) return sample.expected;
  try {
    return resolvePath(sample.expected, path);
  } catch (error) {
    if (error instanceof PathNotFoundError) {
      throw new EvaluatorError(`Expected output has no field '${path}'`);
    }
    throw error;
  }
}

function normalise(text: string, caseSensitive: boolean, collapseWhitespace: boolean): string {
  const collapsed = collapseWhitespace ? pySplit(text).join(" ") : text;
  return caseSensitive ? collapsed : pyCasefold(collapsed);
}

function truncate(text: string, max = 120): string {
  return pyLen(text) <= max ? text : pySlice(text, 0, max - 1) + "…";
}

function isParseOrPathError(error: unknown): boolean {
  return error instanceof JsonParseError || error instanceof PathNotFoundError;
}

// --- exact match -----------------------------------------------------------------------------

export class ExactMatchEvaluator implements Evaluator {
  constructor(private readonly config: Config) {}

  async evaluate(sample: EvaluationSample): Promise<EvaluatorOutcome> {
    const { expected_path, output_path, case_sensitive, collapse_whitespace } = this.config as {
      expected_path: string | null;
      output_path: string | null;
      case_sensitive: boolean;
      collapse_whitespace: boolean;
    };
    const expected = expectedValue(sample, expected_path);
    let actual: unknown = sample.output;
    if (output_path !== null) {
      try {
        actual = resolvePath(parseJsonOutput(sample.output), output_path);
      } catch (error) {
        if (!isParseOrPathError(error)) throw error;
        return binary(false, `Output has no JSON field '${output_path}'`);
      }
    }

    let matched: boolean;
    if (typeof expected === "string" || typeof actual === "string") {
      matched =
        normalise(toText(actual), case_sensitive, collapse_whitespace) ===
        normalise(toText(expected), case_sensitive, collapse_whitespace);
    } else {
      matched = pyEq(actual, expected);
    }
    if (matched) return binary(true, "Output exactly matches the expected value");
    return binary(
      false,
      `Expected ${pyRepr(toText(expected))} but got ${pyRepr(truncate(toText(actual)))}`,
    );
  }
}

// --- contains --------------------------------------------------------------------------------

export class ContainsEvaluator implements Evaluator {
  constructor(private readonly config: Config) {}

  private needles(sample: EvaluationSample): string[] {
    const { values, expected_path } = this.config as {
      values: string[];
      expected_path: string | null;
    };
    const needles = [...values];
    const hasExpected = sample.expected !== null && sample.expected !== undefined;
    if (expected_path !== null || (needles.length === 0 && hasExpected)) {
      const expected = expectedValue(sample, expected_path);
      const items = Array.isArray(expected) ? expected : [expected];
      needles.push(...items.map(toText));
    }
    if (needles.length === 0) {
      throw new EvaluatorError("No values configured and no expected output to search for");
    }
    return needles;
  }

  async evaluate(sample: EvaluationSample): Promise<EvaluatorOutcome> {
    const { mode, case_sensitive } = this.config as { mode: string; case_sensitive: boolean };
    const needles = this.needles(sample);
    const haystack = normalise(sample.output, case_sensitive, true);
    const found = needles.filter((n) => haystack.includes(normalise(n, case_sensitive, true)));
    const missing = needles.filter((n) => !found.includes(n));
    let passed: boolean;
    let score: number;
    if (mode === "any") {
      passed = found.length > 0;
      score = passed ? 1.0 : 0.0;
    } else {
      passed = missing.length === 0;
      score = found.length / needles.length;
    }
    const reason =
      `Found ${found.length} of ${needles.length} expected values` +
      (missing.length ? `; missing ${missing.slice(0, 5).map(pyRepr).join(", ")}` : "");
    return { score, passed, reason, details: { found, missing }, usage: null };
  }
}

// --- regex -----------------------------------------------------------------------------------

/** Translate the Python-only parts of a regular expression into JavaScript syntax. */
export function pythonPatternToJs(pattern: string): string {
  let out = "";
  for (let i = 0; i < pattern.length; i++) {
    const ch = pattern[i];
    if (ch === "\\" && i + 1 < pattern.length) {
      const next = pattern[i + 1];
      if (next === "A") out += "(?<![\\s\\S])";
      else if (next === "Z") out += "(?![\\s\\S])";
      else out += ch + next;
      i++;
    } else if (pattern.startsWith("(?P<", i)) {
      out += "(?<";
      i += 3;
    } else if (pattern.startsWith("(?P=", i)) {
      const end = pattern.indexOf(")", i);
      out += `\\k<${pattern.slice(i + 4, end)}>`;
      i = end;
    } else {
      out += ch;
    }
  }
  return out;
}

export function compilePythonRegex(
  pattern: string,
  flags: { ignoreCase?: boolean; multiline?: boolean; dotall?: boolean } = {},
): RegExp {
  const jsFlags =
    (flags.ignoreCase ? "i" : "") + (flags.multiline ? "m" : "") + (flags.dotall ? "s" : "");
  return new RegExp(pythonPatternToJs(pattern), jsFlags);
}

export class RegexEvaluator implements Evaluator {
  private readonly regex: RegExp;

  constructor(private readonly config: Config) {
    const { pattern, ignore_case, multiline, dotall } = config as {
      pattern: string;
      ignore_case: boolean;
      multiline: boolean;
      dotall: boolean;
    };
    this.regex = compilePythonRegex(pattern, { ignoreCase: ignore_case, multiline, dotall });
  }

  async evaluate(sample: EvaluationSample): Promise<EvaluatorOutcome> {
    const shouldMatch = this.config.should_match as boolean;
    const match = this.regex.exec(sample.output);
    const passed = (match !== null) === shouldMatch;
    let reason = match !== null ? `Pattern matched ${pyRepr(match[0])}` : "Pattern did not match";
    if (!shouldMatch) reason += " (expected no match)";
    return binary(passed, reason, { match: match !== null ? match[0] : null });
  }
}

// --- JSON validity and schema ----------------------------------------------------------------

export class JsonValidEvaluator implements Evaluator {
  constructor(private readonly config: Config) {}

  async evaluate(sample: EvaluationSample): Promise<EvaluatorOutcome> {
    try {
      parseJsonOutput(sample.output, this.config.allow_code_fence as boolean);
    } catch (error) {
      if (error instanceof JsonParseError) return binary(false, error.message);
      throw error;
    }
    return binary(true, "Output is valid JSON");
  }
}

export class JsonSchemaEvaluator implements Evaluator {
  private readonly validator: SchemaValidator;

  constructor(private readonly config: Config) {
    this.validator = new SchemaValidator(config.json_schema as Record<string, unknown>);
  }

  async evaluate(sample: EvaluationSample): Promise<EvaluatorOutcome> {
    let value: unknown;
    try {
      value = parseJsonOutput(sample.output, this.config.allow_code_fence as boolean);
    } catch (error) {
      if (error instanceof JsonParseError) return binary(false, error.message);
      throw error;
    }
    const violations = this.validator.violations(value);
    if (violations.length === 0) return binary(true, "Output conforms to the schema");
    const messages = violations.slice(0, 10);
    return binary(false, `${violations.length} schema violation(s): ${messages[0]}`, {
      violations: messages,
    });
  }
}

// --- required fields -------------------------------------------------------------------------

export class RequiredFieldsEvaluator implements Evaluator {
  constructor(private readonly config: Config) {}

  async evaluate(sample: EvaluationSample): Promise<EvaluatorOutcome> {
    const { fields, allow_null, allow_code_fence } = this.config as {
      fields: string[];
      allow_null: boolean;
      allow_code_fence: boolean;
    };
    let value: unknown;
    try {
      value = parseJsonOutput(sample.output, allow_code_fence);
    } catch (error) {
      if (!(error instanceof JsonParseError)) throw error;
      return {
        score: 0.0,
        passed: false,
        reason: error.message,
        details: { missing: fields },
        usage: null,
      };
    }
    const missing: string[] = [];
    for (const path of fields) {
      let found: unknown;
      try {
        found = resolvePath(value, path);
      } catch (error) {
        if (!(error instanceof PathNotFoundError)) throw error;
        missing.push(path);
        continue;
      }
      if (found === null && !allow_null) missing.push(path);
    }
    const present = fields.length - missing.length;
    const reason = missing.length
      ? `Missing ${missing.length} of ${fields.length} fields: ${missing.join(", ")}`
      : "All required fields are present";
    return {
      score: present / fields.length,
      passed: missing.length === 0,
      reason,
      details: { missing },
      usage: null,
    };
  }
}

// --- field-level structured comparison -------------------------------------------------------

export class FieldMatchEvaluator implements Evaluator {
  constructor(private readonly config: Config) {}

  private valuesMatch(expected: unknown, actual: unknown): boolean {
    const { numeric_tolerance, case_sensitive } = this.config as {
      numeric_tolerance: number;
      case_sensitive: boolean;
    };
    if (typeof expected === "boolean" || typeof actual === "boolean") return pyEq(expected, actual);
    if (typeof expected === "number" && typeof actual === "number") {
      return numeric_tolerance === 0
        ? pyIsClose(expected, actual)
        : pyIsClose(expected, actual, numeric_tolerance);
    }
    if (typeof expected === "string" && typeof actual === "string") {
      return normalise(expected, case_sensitive, true) === normalise(actual, case_sensitive, true);
    }
    return pyEq(expected, actual);
  }

  async evaluate(sample: EvaluationSample): Promise<EvaluatorOutcome> {
    const expected = expectedValue(sample, null);
    let fields = this.config.fields as string[] | null;
    if (fields === null) {
      if (!isDict(expected)) {
        throw new EvaluatorError("Field match needs an expected JSON object or explicit fields");
      }
      fields = Object.keys(expected);
    }
    if (fields.length === 0) throw new EvaluatorError("No fields to compare");

    let actual: unknown;
    try {
      actual = parseJsonOutput(sample.output, this.config.allow_code_fence as boolean);
    } catch (error) {
      if (!(error instanceof JsonParseError)) throw error;
      return { score: 0.0, passed: false, reason: error.message, details: {}, usage: null };
    }

    const comparisons: Array<Record<string, unknown>> = [];
    for (const path of fields) {
      let expectedValueAtPath: unknown;
      try {
        expectedValueAtPath = resolvePath(expected, path);
      } catch (error) {
        if (!(error instanceof PathNotFoundError)) throw error;
        throw new EvaluatorError(`Expected output has no field '${path}'`);
      }
      let actualValue: unknown;
      let matched: boolean;
      try {
        actualValue = resolvePath(actual, path);
        matched = this.valuesMatch(expectedValueAtPath, actualValue);
      } catch (error) {
        if (!(error instanceof PathNotFoundError)) throw error;
        actualValue = null;
        matched = false;
      }
      comparisons.push({
        field: path,
        expected: expectedValueAtPath,
        actual: actualValue,
        match: matched,
      });
    }

    const matchedCount = comparisons.filter((c) => c.match).length;
    const score = matchedCount / comparisons.length;
    const mismatched = comparisons.filter((c) => !c.match).map((c) => c.field as string);
    const reason = mismatched.length
      ? `${matchedCount} of ${comparisons.length} fields match; mismatched: ${mismatched.join(", ")}`
      : `All ${comparisons.length} fields match`;
    return {
      score,
      passed: score >= (this.config.pass_threshold as number),
      reason,
      details: { fields: comparisons },
      usage: null,
    };
  }
}
